### PR TITLE
Redesign feature flag management

### PR DESCRIPTION
**Pull request by liry**
Lets move support of feature flags of all scopes to it's own service. Also fix naming conventions and use DTO objects with value added as return types.

### CODE REVIEW

**Comment by liry** [comment]
relates #195

**Comment by liry** [comment]
Rebased and AT fixed

**Comment by liry** [comment]
rebased

File: src/main/java/com/gooddata/featureflag/ProjectFeatureFlag.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    @JsonIgnore
+    private Links links;
+
+    /**
+     * Creates new project feature flag which is by default enabled (true).
+     *
+     * @param name unique name of feature flag
+     */
+    public ProjectFeatureFlag(String name) {
+        this(name, true, null);
```
Comment (martiner) [inline src/main/java/com/gooddata/featureflag/ProjectFeatureFlag.java]: check name?

File: src/main/java/com/gooddata/featureflag/ProjectFeatureFlag.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    /**
+     * Creates new project feature flag with given value.
+     *
+     * @param name unique name of feature flag
+     * @param enabled true (flag enabled) or false (flag disabled)
+     */
+    public ProjectFeatureFlag(String name, boolean enabled) {
+        this(name, enabled, null);
```
Comment (martiner) [inline src/main/java/com/gooddata/featureflag/ProjectFeatureFlag.java]: check name?

File: src/main/java/com/gooddata/featureflag/FeatureFlag.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    public FeatureFlag(final String name, final boolean enabled) {
+        this.name = notNull(name, "name cannot be null");
+        this.enabled = enabled;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public boolean getEnabled() {
```
Comment (martiner) [inline src/main/java/com/gooddata/featureflag/FeatureFlag.java]: isEnabled()?

File: src/main/java/com/gooddata/featureflag/ProjectFeatureFlag.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        this.links = links;
+    }
+
+    @JsonProperty("key")
+    public String getName() {
+        return name;
+    }
+
+    @JsonProperty("value")
+    public boolean getEnabled() {
```
Comment (martiner) [inline src/main/java/com/gooddata/featureflag/ProjectFeatureFlag.java]: isEnabled()?

File: src/main/java/com/gooddata/featureflag/FeatureFlagService.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+    /**
+     * Returns aggregated feature flags for given project and current user (aggregates global, project group, project
+     * and user feature flags).
+     * It doesn't matter whether feature flag is enabled or not, it'll be included in both cases.
+     *
+     * @param project project, cannot be null
+     * @return aggregated feature flags for given project and current user
+     */
+    public FeatureFlags getFeatureFlags(final Project project) {
```
Comment (martiner) [inline src/main/java/com/gooddata/featureflag/FeatureFlagService.java]: I tend to name `listFeatureFlags()` here:
# the method used to be called `listAggregatedFeatureFlags()`
# we name listing methods with `list` prefix (eg. WarehouseService.listWarehouses(), ProcessService.listProcesses(),...) and the `FeatureFlags` class is more like collection

File: src/main/java/com/gooddata/featureflag/FeatureFlagService.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+    /**
+     * Returns project feature flags (only project scoped flags, use {@link #getFeatureFlags(Project)} for aggregated
+     * flags from all scopes) for given project.
+     * It doesn't matter whether feature flag is enabled or not, it'll be included in both cases.
+     *
+     * @param project project, cannot be null
+     * @return list of all feature flags for given project
+     */
+    public ProjectFeatureFlags getProjectFeatureFlags(final Project project) {
```
Comment (martiner) [inline src/main/java/com/gooddata/featureflag/FeatureFlagService.java]: listFeatureFlags()?

File: src/main/java/com/gooddata/featureflag/FeatureFlags.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+import java.util.LinkedList;
+import java.util.List;
+
+import static com.gooddata.util.Validate.notEmpty;
+import static org.springframework.util.Assert.notNull;
+
+@JsonTypeInfo(include = JsonTypeInfo.As.WRAPPER_OBJECT, use = JsonTypeInfo.Id.NAME)
+@JsonTypeName("featureFlags")
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class FeatureFlags implements Iterable<FeatureFlag> {
```
Comment (martiner) [inline src/main/java/com/gooddata/featureflag/FeatureFlags.java]: iterable is cool, are we sure we don't want the real collection? (similarly we have PageableCollection)
i guess we had this conversation before, don't you remember my opinion? :)

File: src/main/java/com/gooddata/featureflag/FeatureFlagService.java (hunk 6 above)
Comment (liry) [inline src/main/java/com/gooddata/featureflag/FeatureFlagService.java]: listProjectFeatureFlags?

File: src/main/java/com/gooddata/featureflag/FeatureFlags.java (hunk 7 above)
Comment (liry) [inline src/main/java/com/gooddata/featureflag/FeatureFlags.java]: I don't remember it, but I'll change ti to List the same way as PageableList ;)

**Pull request merged**